fedscale_examples/Async_FL/async_aggregator.py

```diff
@@ -0,0 +1,489 @@
+        }
+        return conf
+
+    def create_client_task(self, executorId):
+        """Issue a new client training task to the executor"""
+
+        next_clientId = self.resource_manager.get_next_task(executorId)
+        train_config = None
+        # NOTE: model = None then the executor will load the global model broadcasted in UPDATE_MODEL
+        model = None
+        # TODO: useless model?
+        if next_clientId != None:
+            config = self.get_client_conf(next_clientId)
+            train_config = {'client_id': next_clientId, 'task_config': config}
+        return train_config, model
+
+    def CLIENT_REGISTER(self, request, context):
+        """FL Client register to the aggregator"""
+
+        # NOTE: client_id = executor_id in deployment,
+        # while multiple client_id uses the same executor_id (VMs) in simulations
+        executor_id = request.executor_id
+        executor_info = self.deserialize_response(request.executor_info)
+        if executor_id not in self.individual_client_events:
+            # logging.info(f"Detect new client: {executor_id}, executor info: {executor_info}")
+            self.individual_client_events[executor_id] = collections.deque()
+        else:
+            logging.info(f"Previous client: {executor_id} resumes connecting")
+
+        # We can customize whether to admit the clients here
+        self.executor_info_handler(executor_id, executor_info)
+        dummy_data = self.serialize_response(events.DUMMY_RESPONSE)
+
+        return job_api_pb2.ServerResponse(event=events.DUMMY_EVENT,
+                                          meta=dummy_data, data=dummy_data)
+
+    def CLIENT_PING(self, request, context):
+        """Handle client requests"""
+
+        # NOTE: client_id = executor_id in deployment,
+        # while multiple client_id may use the same executor_id (VMs) in simulations
+        executor_id, client_id = request.executor_id, request.client_id
+        response_data = response_msg = events.DUMMY_RESPONSE
+
+        if len(self.individual_client_events[executor_id]) == 0:
+            # send dummy response
+            current_event = events.DUMMY_EVENT
+            response_data = response_msg = events.DUMMY_RESPONSE
+        else:
+            current_event = self.individual_client_events[executor_id].popleft()
+            if current_event == events.CLIENT_TRAIN:
+                response_msg, response_data = self.create_client_task(client_id)
+                if response_msg is None:
+                    current_event = events.DUMMY_EVENT
+                    if self.experiment_mode != events.SIMULATION_MODE:
+                        self.individual_client_events[executor_id].appendleft(events.CLIENT_TRAIN)
+            elif current_event == events.MODEL_TEST:
+                response_msg = self.get_test_config(client_id)
+            elif current_event == events.UPDATE_MODEL:
+                response_data = self.get_global_model()
```

Comment:
For my understanding, is the model serialized as a pickle into bytes and then passed through grpc?

Reply:
yes !

Reply:
Thanks! To better emulate what an Android or iOS device sends back post-training, I think the data could be of type bytes[] for better cross platform compatibility. Where each layer's params are represented as a single bytes object. This is how Flower serializes their models.

Reply:
Thanks for your feedback. We will enable this in our next release soon.